### PR TITLE
merging braille-printer-application with cups-filters code 

### PR DESCRIPTION
**Pull request by chandresh-soni**
@tillkamppeter could you review it and let me know if any changes are required

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -5,7 +5,7 @@ doc_DATA = \
 	AUTHORS \
 	COPYING \
 	CHANGES.md \
-	CHANGES-1.x.md \
+	CHANGES-cups-filter.md \
```
Comment (tillkamppeter) [inline Makefile.am:8]: `CHANGES-cups-filters.md` it is cups-filter**s**, not cups-filter.

File: braille-printer-app/LICENCE [hunk 2]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (tillkamppeter) [inline braille-printer-app/LICENCE:1]: You can remove LICENSE and NOTICE from this subdir. they are already in the main directory.

File: braille-printer-app/NOTICE [hunk 3]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (tillkamppeter) [inline braille-printer-app/NOTICE:1]: Please update the copyright/author name part of `NOTICE` in the main directory. Do not add an extra `NOTICE` file here.

File: braille-printer-app/readme.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
+Braille Printer Application
```
Comment (tillkamppeter) [inline braille-printer-app/readme.md:1]: No extra README.md in this sub-directory needed. Merge with the main directory's README.md. Let it have a section for the Printer Application and a section for the classic driver, the Printer Application, being the default/preferred method.

File: braille-printer-app/brf-printer-app.c [hunk 5]
```diff
@@ -0,0 +1,821 @@
...
+typedef struct brf_spooling_conversion_s
+{
+  char *srctype;                   // Input data type
+  char *dsttype;                   // Output data type
+  int num_filters;                       // Number of filters
+  cf_filter_filter_in_chain_t filters[]; // List of filters with
+                                         // parameters
+} brf_spooling_conversion_t;
+
+static brf_spooling_conversion_t brf_convert_pdf_to_brf =
```
Comment (tillkamppeter) [inline braille-printer-app/brf-printer-app.c:530]: - Does this actually convert **PDF** to BRF or is this converting text to BRF?
- This structure `brf_convert_pdf_to_brf` is nowhere used in the code. So do we really need it?

File: braille-printer-app/brf-printer-app.c [hunk 6]
```diff
@@ -0,0 +1,821 @@
...
+  filter_data_ext->envp= NULL;
+
+
+  //
+  // Find filters to use for this job
+  //
+
+  for (conversion =
+           (brf_spooling_conversion_t *)
+               cupsArrayFirst(spooling_conversions);
```
Comment (tillkamppeter) [inline braille-printer-app/brf-printer-app.c:650]: Where did `spooling_conversions` get populated?

File: braille-printer-app/brf-printer-app.c [hunk 7]
```diff
@@ -0,0 +1,821 @@
...
+  // Set up filter function chain
+  //
+
+  chain = cupsArrayNew(NULL, NULL);
+
+  for (int i = 0; i < conversion->num_filters; i++){
+    cupsArrayAdd(chain, &(conversion->filters[i]));
+    cupsArrayAdd(chain,&(filter_data_ext));
+    }
+
```
Comment (tillkamppeter) [inline braille-printer-app/brf-printer-app.c:696]: The filter chain `chain` contains conversion filters from `conversion`. These must be of type `cf_filter_filter_in_chain_t`, but `filter_data_ext` is of type `cf_filter_external_t`. Also does it have to be added to `chain` after each of the conversion filters?

File: braille-printer-app/brf-printer-app.c [hunk 8]
```diff
@@ -0,0 +1,821 @@
...
+      make = cupsGetOption("MFG", num_did, did);
+
+  
+  // // Then loop through the driver list to find the best match...
+  // for (i = 0; i < (int)(sizeof(brf_drivers) / sizeof(brf_drivers[0])); i ++)
+  {
+ 
+
+    if (brf_drivers[i].device_id)
+    {
```
Comment (tillkamppeter) [inline braille-printer-app/brf-printer-app.c:104]: brf_drivers[] has only 1 entry without device ID, so `autoadd_cb()` always returns NULL. Did you test the code? Or only compile it?

File: braille-printer-app/brf-printer-app.c [hunk 9]
```diff
@@ -0,0 +1,821 @@
...
+      }
+    }
+    else
+    {
+      // No match
+      score = 0;
+      break;
+    }
+  }
+}
```
Comment (tillkamppeter) [inline braille-printer-app/brf-printer-app.c:187]: You need to return the core here.

File: braille-printer-app/brf-printer-app.c [hunk 10]
```diff
@@ -0,0 +1,821 @@
...
+  {
+    if (!strcmp(driver_name, brf_drivers[i].name))
+    {
+      papplCopyString(data->make_and_model, brf_drivers[i].description, sizeof(data->make_and_model));
+      break;
+    }
+  }
+
+  // Pages per minute 
+  data->ppm = 60;
```
Comment (tillkamppeter) [inline braille-printer-app/brf-printer-app.c:217]: Braille embossers seem to be really fast ...

File: braille-printer-app/brf-printer-app.c [hunk 11]
```diff
@@ -0,0 +1,821 @@
...
+
+ 
+  // Color values...
+  data->color_supported   = PAPPL_COLOR_MODE_AUTO | PAPPL_COLOR_MODE_MONOCHROME;
+  data->color_default   = PAPPL_COLOR_MODE_MONOCHROME;
+  data->raster_types  = PAPPL_PWG_RASTER_TYPE_BLACK_8; // to be done just guess
```
Comment (tillkamppeter) [inline braille-printer-app/brf-printer-app.c:223]: I raster input really 8-bit grayscale? And not 1-bit bi-level? (Depends on the filters)

File: braille-printer-app/brf-printer-app.c (hunk 10 above)
Comment (sthibaul) [reply braille-printer-app/brf-printer-app.c:217]: Heh :)
Indeed it's rather `1` that should be set. Yes, it typically takes one minute to emboss a whole page...

**Comment by sthibaul** [comment]
Superseded by https://github.com/OpenPrinting/braille-printer-app/pull/10

**Pull request closed without merging**